inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx

```diff
@@ -1,80 +1,57 @@
+import type { PageProps } from "./index.page.jsx"
 
-/**
- * The root of the repository.
- *
- * Makes it possible to use absolute paths for rendering markdown.
- */
 const repositoryRoot = new URL("../../../../../../", import.meta.url)
 
-/**
- * the table of contents without the html for each document
- * saving bandwith and speeding up the site)
- */
-export type ProcessedTableOfContents = Record<
+export type GeneratedTableOfContents = Record<
 	string,
-	Awaited<ReturnType<typeof parseMarkdown>>["frontmatter"]
+	{
+		title: string
+		description: string
+		href: string
+	}
 >
 
-/**
- * The index of documentation markdown files. The href's acts as id.
- *
- * @example
- * 	{
- * 		"/documentation/intro": document,
- * 	}
- */
-const index: Record<string, Awaited<ReturnType<typeof parseMarkdown>>> = {}
-/**
- * the table of contents without the html for each document
- * saving bandwith and speeding up the site)
- */
-const processedTableOfContents: ProcessedTableOfContents = {}
+const index: Record<string, Awaited<ReturnType<any>>> = {}
+const generatedTableOfContents = {} as GeneratedTableOfContents
 
 await generateIndexAndTableOfContents()
 
-// should only run server side
 export const onBeforeRender: OnBeforeRender<PageProps> = async (pageContext) => {
-	// dirty way to get reload of markdown (not hot reload though)
 	if (import.meta.env.DEV) {
 		await generateIndexAndTableOfContents()
 	}
 	if (
-		!Object.keys(index).includes(pageContext.urlPathname) &&
+		!Object.keys(index).includes(pageContext.urlPathname.replace("/blog/", "")) &&
 		pageContext.urlPathname !== "/blog"
 	) {
 		throw RenderErrorPage({ pageContext: { is404: true } })
 	}
+
 	return {
 		pageContext: {
 			pageProps: {
-				markdown: index[pageContext.urlPathname]!,
-				processedTableOfContents: processedTableOfContents,
+				markdown: index[pageContext.urlPathname.replace("/blog/", "")]!,
+				meta: generatedTableOfContents[pageContext.urlPathname.replace("/blog/", "")]!,
+				processedTableOfContents: generatedTableOfContents,
 			},
 		},
 	}
 }
 
-/**
- * Generates the index and table of contents.
- */
 async function generateIndexAndTableOfContents() {
-	for (const path of tableOfContents) {
-		const text = await fs.readFile(new URL(`inlang/blog/${path}`, repositoryRoot), "utf-8")
-		const markdown = parseMarkdown({
-			text,
-			frontmatterSchema: BlogFrontmatterSchema,
-		})
-		// not pushing to processedTableOfContents directly in case
-		// the category is undefined so far
-		processedTableOfContents[markdown.frontmatter.href] = markdown.frontmatter
+	for (const content of tableOfContents) {
+		const raw = await fs.readFile(new URL(`blog/${content.path}`, repositoryRoot), "utf-8")
+		const markdown = await convert(raw)
 
-		index[markdown.frontmatter.href] = markdown
+		index[content.slug] = markdown
+		generatedTableOfContents[content.slug] = {
+			...content,
+			href: "/blog/" + content.slug,
+		}
 	}
 }
```

Comment:
You don't need to generate the table of contents anymore. We already have it in the form of the JSON. 

**What to change**

1. create a variable `renderedMarkdown` at the root level which will store already rendered markdown.
2. In `onBeforeRender()`, check if the requested blog is contained in `renderedMarkdown`. Otherwise, render it and store the result in `renderedMarkdown`. 
3. Remove the `generateIndexAndTableOfContents()` function. 
4. Remove `processedTableOfContents` from the page context as the client can directly import the json
5. REPEAT IN DOCUMENTATION SITE

```ts
const renderedMarkdown = {}

export async function onBeforeRender(pageContext){
  const slug = pageContext.urlPathname.slice(`/blog/`.length)
  if (renderedMarkdown[slug] === undefined){
    // pseudocode
    const content = tableOfContents.find(slug)
    const text = await fs.readFile(new URL(`blog/${content.path}`, repositoryRoot), "utf-8")
    const markdown = await convert(text)
    renderedMarkdown[slug] = markdown
  }
  return {
    pageContext: {
      pageProps: {
        markdown: renderedMarkdown[slug]
      }
    }
  }
}
```

Reply:
The code above is all we need to render markdown :)